File: example-objc/example-objc/CardFormViewScrollViewController.m
//
//  CardFormViewScrollViewController.m
//  example-objc
//
//  Created by Tadashi Wakayanagi on 2019/08/28.
//  Copyright © 2019 PAY, Inc. All rights reserved.
//

#import "CardFormViewScrollViewController.h"
#import "ColorTheme.h"
#import "UIViewController+Alert.h"
@import PAYJP;

@interface CardFormViewScrollViewController ()

@property(weak, nonatomic) IBOutlet PAYCardFormLabelStyledView *cardFormView;
@property(weak, nonatomic) IBOutlet UIButton *createTokenButton;
@property(weak, nonatomic) IBOutlet UIButton *validateAndCreateTokenButton;
@property(weak, nonatomic) IBOutlet UITextField *selectColorField;
@property(weak, nonatomic) IBOutlet UILabel *tokenIdLabel;

- (IBAction)createToken:(id)sender;
- (IBAction)validateAndCreateToken:(id)sender;

@property(strong, nonatomic) NSArray *list;
@property(strong, nonatomic) UIPickerView *pickerView;
@property(assign, nonatomic) PAYTokenOperationStatus tokenOperationStatus;
@property(nonatomic, strong) PAYToken *pendingToken;

@end

@implementation CardFormViewScrollViewController

- (void)viewDidLoad {
  [super viewDidLoad];

  self.cardFormView.delegate = self;
  [self fetchBrands];

  self.list = @[ @"Normal", @"Red", @"Blue", @"Dark" ];
  self.pickerView = [[UIPickerView alloc] init];
  self.pickerView.delegate = self;
  self.pickerView.dataSource = self;
  self.pickerView.showsSelectionIndicator = YES;
  self.selectColorField.delegate = self;

  UIToolbar *toolbar = [[UIToolbar alloc] init];
  UIBarButtonItem *spaceItem =
      [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace
                                                    target:self
                                                    action:nil];
  UIBarButtonItem *doneItem =
      [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone
                                                    target:self
                                                    action:@selector(colorSelected:)];
  [toolbar setItems:@[ spaceItem, doneItem ]];
  [toolbar sizeToFit];

  self.selectColorField.inputView = self.pickerView;
  self.selectColorField.inputAccessoryView = toolbar;

  [[NSNotificationCenter defaultCenter] addObserver:self
                                           selector:@selector(handleTokenOperationStatusChange:)
                                               name:NSNotification.payjpTokenOperationStatusChanged
                                             object:nil];
}

- (void)colorSelected:(id)sender {
  [self.selectColorField endEditing:YES];
  NSString *selected = self.list[[self.pickerView selectedRowInComponent:0]];
  self.selectColorField.text = selected;
  ColorTheme theme = GetColorTheme(selected);

  switch (theme) {
    case Red: {
      UIColor *red = RGB(255, 69, 0);
      PAYCardFormStyle *style = [[PAYCardFormStyle alloc] initWithLabelTextColor:red
                                                                  inputTextColor:red
                                                                  errorTextColor:nil
                                                                       tintColor:red
                                                       inputFieldBackgroundColor:nil
                                                               submitButtonColor:nil
                                                                  highlightColor:nil];
      [self.cardFormView applyWithStyle:style];
      self.cardFormView.backgroundColor = UIColor.clearColor;
      break;
    }
    case Blue: {
      UIColor *blue = RGB(0, 103, 187);
      PAYCardFormStyle *style = [[PAYCardFormStyle alloc] initWithLabelTextColor:blue
                                                                  inputTextColor:blue
                                                                  errorTextColor:nil
                                                                       tintColor:blue
                                                       inputFieldBackgroundColor:nil
                                                               submitButtonColor:nil
                                                                  highlightColor:nil];
      [self.cardFormView applyWithStyle:style];
      self.cardFormView.backgroundColor = UIColor.clearColor;
      break;
    }
    case Dark: {
      UIColor *white = UIColor.whiteColor;
      UIColor *darkGray = RGB(61, 61, 61);
      UIColor *lightGray = RGB(80, 80, 80);
      PAYCardFormStyle *style = [[PAYCardFormStyle alloc] initWithLabelTextColor:white
                                                                  inputTextColor:white
                                                                  errorTextColor:nil
                                                                       tintColor:white
                                                       inputFieldBackgroundColor:lightGray
                                                               submitButtonColor:nil
                                                                  highlightColor:nil];
      [self.cardFormView applyWithStyle:style];
      self.cardFormView.backgroundColor = darkGray;
      break;
    }
    default: {
      UIColor *black = UIColor.blackColor;
      UIColor *defaultBlue = RGB(0, 122, 255);
      PAYCardFormStyle *style = [[PAYCardFormStyle alloc] initWithLabelTextColor:black
                                                                  inputTextColor:black
                                                                  errorTextColor:nil
                                                                       tintColor:defaultBlue
                                                       inputFieldBackgroundColor:nil
                                                               submitButtonColor:nil
                                                                  highlightColor:nil];
      [self.cardFormView applyWithStyle:style];
      self.cardFormView.backgroundColor = UIColor.clearColor;
      break;
    }
  }
}

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
  return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
  return self.list.count;
}

- (NSString *)pickerView:(UIPickerView *)pickerView
             titleForRow:(NSInteger)row
            forComponent:(NSInteger)component {
  return self.list[row];
}

- (BOOL)textField:(UITextField *)textField
    shouldChangeCharactersInRange:(NSRange)range
                replacementString:(NSString *)string {
  return NO;
}

- (void)formInputValidatedIn:(UIView *)cardFormView isValid:(BOOL)isValid {
  [self updateButtonEnabled];
}

- (void)formInputDoneTappedIn:(UIView *)cardFormView {
  [self createToken];
}

- (void)handleTokenOperationStatusChange:(NSNotification *)notification {
  self.tokenOperationStatus =
      [notification.userInfo[PAYNotificationKey.newTokenOperationStatus] integerValue];
  [self updateButtonEnabled];
}

- (void)updateButtonEnabled {
  BOOL isAcceptable = self.tokenOperationStatus == PAYTokenOperationStatusAcceptable;
  self.createTokenButton.enabled = self.cardFormView.isValid && isAcceptable;
  self.validateAndCreateTokenButton.enabled = isAcceptable;
}

- (IBAction)createToken:(id)sender {
  if (!self.cardFormView.isValid) {
    return;
  }
  [self createToken];
}

- (IBAction)validateAndCreateToken:(id)sender {
  BOOL isValid = [self.cardFormView validateCardForm];
  if (isValid) {
    [self createToken];
  }
}

- (void)createToken {
  __weak typeof(self) wself = self;

  [self.cardFormView
      createTokenWith:nil
      useThreeDSecure:YES
           completion:^(PAYToken *token, NSError *error) {
             if (error.domain == PAYErrorDomain && error.code == PAYErrorServiceError) {
               id<PAYErrorResponseType> errorResponse = error.userInfo[PAYErrorServiceErrorObject];
               NSLog(@"[errorResponse] %@", errorResponse.description);
             }

             if (!token) {
               dispatch_async(dispatch_get_main_queue(), ^{
                 wself.tokenIdLabel.text = nil;
                 [wself showError:error];
               });
               return;
             }

             NSLog(@"token = %@", [wself displayToken:token]);
             dispatch_async(dispatch_get_main_queue(), ^{
               if (token.card.threeDSecureStatus == PAYThreeDSecureStatusUnverified) {
                 wself.pendingToken = token;
                 [[PAYJPThreeDSecureProcessHandler sharedHandler]
                     startThreeDSecureProcessWithViewController:wself
                                                       delegate:wself
                                                     resourceId:token.identifer];
                 return;
               }
               wself.tokenIdLabel.text = token.identifer;
               [wself showToken:token];
             });
           }];
}

- (void)fetchBrands {
  __weak typeof(self) wself = self;

  [self.cardFormView
      fetchBrandsWith:@"tenant_id"
           completion:^(NSArray<NSString *> *cardBrands, NSError *error) {
             if (error.domain == PAYErrorDomain && error.code == PAYErrorServiceError) {
               id<PAYErrorResponseType> errorResponse = error.userInfo[PAYErrorServiceErrorObject];
               NSLog(@"[errorResponse] %@", errorResponse.description);
             }

             if (!cardBrands) {
               dispatch_async(dispatch_get_main_queue(), ^{
                 [wself showError:error];
               });
             }
           }];
}

- (void)completeTokenTds {
  if (!self.pendingToken) {
    return;
  }

  __weak typeof(self) wself = self;
  [[PAYAPIClient sharedClient]
      finishTokenThreeDSecureWith:self.pendingToken.identifer
                completionHandler:^(PAYToken *token, NSError *error) {
                  if (error) {
                    if ([error.domain isEqualToString:PAYErrorDomain]) {
                      id<PAYErrorResponseType> errorResponse =
                          error.userInfo[PAYErrorServiceErrorObject];
                      NSLog(@"[errorResponse] %@", errorResponse.description);
                    }

                    dispatch_async(dispatch_get_main_queue(), ^{
                      wself.tokenIdLabel.text = nil;
                      [wself showError:error];
                    });
                    return;
                  }

                  dispatch_async(dispatch_get_main_queue(), ^{
                    wself.pendingToken = nil;
                    wself.tokenIdLabel.text = token.identifer;
                    [wself showToken:token];
                  });
                }];
}

#pragma mark - PAYThreeDSecureProcessHandlerDelegate

- (void)threeDSecureProcessHandlerDidFinish:(PAYJPThreeDSecureProcessHandler *)handler
                                     status:(enum ThreeDSecureProcessStatus)status {
  switch (status) {
    case ThreeDSecureProcessStatusCompleted:
      [self completeTokenTds];
      break;
    case ThreeDSecureProcessStatusCanceled:
      // UI更新など
      break;
    default:
      break;
  }
}

@end

File: example-swift/example-swift/CardFormViewScrollViewController.swift
//
//  CardFormViewScrollViewController.swift
//  example-swift
//
//  Created by Tadashi Wakayanagi on 2019/09/12.
//

import PAYJP

class CardFormViewScrollViewController: UIViewController, CardFormViewDelegate,
                                        UIPickerViewDelegate, UIPickerViewDataSource, UITextFieldDelegate {

    @IBOutlet private weak var formContentView: UIView!
    @IBOutlet private weak var createTokenButton: UIButton!
    @IBOutlet private weak var validateAndCreateTokenButton: UIButton!
    @IBOutlet private weak var tokenIdLabel: UILabel!
    @IBOutlet private weak var selectColorField: UITextField!

    private var cardFormView: CardFormLabelStyledView!
    private var tokenOperationStatus: TokenOperationStatus = .acceptable

    private let list: [ColorTheme] = [.Normal, .Red, .Blue, .Dark]
    private var pickerView: UIPickerView!

    override func viewDidLoad() {
        // Carthageを使用している関係でstoryboardでCardFormViewを指定できないため
        // storyboardに設置しているViewにaddSubviewする形で実装している
        let x: CGFloat = self.formContentView.bounds.origin.x
        let y: CGFloat = self.formContentView.bounds.origin.y

        let width: CGFloat = self.formContentView.bounds.width
        let height: CGFloat = self.formContentView.bounds.height

        let frame: CGRect = CGRect(x: x, y: y, width: width, height: height)
        cardFormView = CardFormLabelStyledView(frame: frame)
        cardFormView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        cardFormView.delegate = self

        self.formContentView.addSubview(cardFormView)

        self.pickerView = UIPickerView()
        self.pickerView.delegate = self
        self.pickerView.dataSource = self
        self.pickerView.showsSelectionIndicator = true
        self.selectColorField.delegate = self

        let toolbar = UIToolbar()
        let spaceItem = UIBarButtonItem(
            barButtonSystemItem: UIBarButtonItem.SystemItem.flexibleSpace,
            target: nil,
            action: nil)
        let doneItem = UIBarButtonItem(
            barButtonSystemItem: UIBarButtonItem.SystemItem.done,
            target: self,
            action: #selector(colorSelected(_:)))
        toolbar.setItems([spaceItem, doneItem], animated: true)
        toolbar.sizeToFit()

        self.selectColorField.inputView = self.pickerView
        self.selectColorField.inputAccessoryView = toolbar

        self.fetchBrands()

        NotificationCenter.default.addObserver(self,
                                               selector: #selector(handleTokenOperationStatusChange(notification:)),
                                               name: .payjpTokenOperationStatusChanged,
                                               object: nil)
    }

    @objc private func colorSelected(_ sender: UIButton) {
        self.selectColorField.endEditing(true)
        let theme = self.list[self.pickerView.selectedRow(inComponent: 0)]
        self.selectColorField.text = theme.rawValue

        switch theme {
        case .Red:
            let red = UIColor(255, 69, 0)
            let style = FormStyle(
                labelTextColor: red,
                inputTextColor: red,
                tintColor: red)
            self.cardFormView.apply(style: style)
            self.cardFormView.backgroundColor = .clear
        case .Blue:
            let blue = UIColor(0, 103, 187)
            let style = FormStyle(
                labelTextColor: blue,
                inputTextColor: blue,
                tintColor: blue)
            self.cardFormView.apply(style: style)
            self.cardFormView.backgroundColor = .clear
        case .Dark:
            let darkGray = UIColor(61, 61, 61)
            let lightGray = UIColor(80, 80, 80)
            let style = FormStyle(
                labelTextColor: .white,
                inputTextColor: .white,
                tintColor: .white,
                inputFieldBackgroundColor: lightGray)
            self.cardFormView.apply(style: style)
            self.cardFormView.backgroundColor = darkGray
        default:
            let defaultBlue = UIColor(0, 122, 255)
            let style = FormStyle(
                labelTextColor: .black,
                inputTextColor: .black,
                tintColor: defaultBlue)
            self.cardFormView.apply(style: style)
            self.cardFormView.backgroundColor = .clear
        }
    }

    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }

    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return self.list.count
    }

    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        return self.list[row].rawValue
    }

    func textField(
        _ textField: UITextField,
        shouldChangeCharactersIn range: NSRange,
        replacementString string: String
    ) -> Bool {
        return false
    }

    func formInputValidated(in cardFormView: UIView, isValid: Bool) {
        self.updateButtonEnabled()
    }

    func formInputDoneTapped(in cardFormView: UIView) {
        self.createToken()
    }

    @objc private func handleTokenOperationStatusChange(notification: Notification) {
        if let value = notification.userInfo?[PAYNotificationKey.newTokenOperationStatus] as? Int,
           let newStatus = TokenOperationStatus.init(rawValue: value) {
            self.tokenOperationStatus = newStatus
            self.updateButtonEnabled()
        }
    }

    private func updateButtonEnabled() {
        let isAcceptable = self.tokenOperationStatus == .acceptable
        self.createTokenButton.isEnabled = isAcceptable && self.cardFormView.isValid
        self.validateAndCreateTokenButton.isEnabled = isAcceptable
    }

    @IBAction func createToken(_ sender: Any) {
        if !self.cardFormView.isValid {
            return
        }
        createToken()
    }

    @IBAction func validateAndCreateToken(_ sender: Any) {
        let isValid = self.cardFormView.validateCardForm()
        if isValid {
            createToken()
        }
    }

    func createToken() {
        self.cardFormView.createToken(tenantId: "tenant_id") { [weak self] result in
            guard let self = self else { return }
            switch result {
            case .success(let token):
                DispatchQueue.main.async {
                    self.tokenIdLabel.text = token.identifer
                    self.showToken(token: token)
                }
            case .failure(let error):
                if let apiError = error as? APIError, let payError = apiError.payError {
                    print("[errorResponse] \(payError.description)")
                }

                DispatchQueue.main.async {
                    self.tokenIdLabel.text = nil
                    self.showError(error: error)
                }
            }
        }
    }

    func fetchBrands() {
        self.cardFormView.fetchBrands(tenantId: "tenant_id") { [weak self] result in
            guard let self = self else { return }
            switch result {
            case .success(let brands):
                print("card brands => \(brands)")
            case .failure(let error):
                if let payError = error.payError {
                    print("[errorResponse] \(payError.description)")
                }

                DispatchQueue.main.async {
                    self.tokenIdLabel.text = nil
                    self.showError(error: error)
                }
            }
        }
    }
}
